### PR TITLE
Add support for hermetic nixos configurations

### PR DESCRIPTION
**Pull request by ixmatus**
This change adds a Terraform variable `hermetic` and supporting logic to the Nix expression evaluated by `nix-instantiate` to better support users that need to evaluate configurations that supply their own pinned nixpkgs or wrap eval-modules to customize module special args.

e.g. at Awake we pin nixpkgs and build our NixOS configuration closures completely sealed from anything on the NIX_PATH. This change simply allows one to say "I'm supplying my own expression that results in a nixos system, so please use that".

A minor additional change was to further purify the Nix expression by using the supplied `pkgs.lib` from the derived NixOS system instead of effectfully re-importing <nixpkgs>.

### CODE REVIEW

File: deploy_nixos/nixos-instantiate.sh [hunk 1]
```diff
@@ -8,17 +8,31 @@ config_pwd=$3
...
+  { system, configuration, hermetic ? false, ... }:
   let
-    os = import <nixpkgs/nixos> { inherit system configuration; };
-    inherit (import <nixpkgs/lib>) concatStringsSep;
+    os =
+      if hermetic
+        then import configuration
+        else import <nixpkgs/nixos> { inherit system configuration; };
+
+    inherit (os.pkgs) lib;
```
Comment (zimbatm) [inline deploy_nixos/nixos-instantiate.sh:23]: Using `builtins.concatStringsSep` might be enough since it's the only lib function that is used below.

**Review by zimbatm — CHANGES_REQUESTED** [review]
Good idea!

Do you mind also giving an example on how you make the configuration hermetic?

File: deploy_nixos/nixos-instantiate.sh (hunk 1 above)
Comment (ixmatus) [reply deploy_nixos/nixos-instantiate.sh:23]: TIL that concatStringsSep is a built-in! That's definitely preferred IMHO.

**Comment by ixmatus** [comment]
@zimbatm does this meet your review requirements?

**Comment by zimbatm** [comment]
thanks!

**Pull request merged**